[3D] Support transparent entities in the camera controller

ptitjano opened the pull request:
## Description

When clicked on a 3d scene to make a rotation for example, `QgsCameraController` needs to compute 3d coordinates from 2d mouse coordinates. It uses the depth buffer in order to estimate the distance between the camera position and the clicked position. This approach works very well as long as the objects of the scene are not
transparent. Indeed, transparent objects do not appear in the depth buffer. This means that `QgsCameraController` does not work properly on a 3d scene which contains transparent vector layers.

At the moment, this is not really a problem because most of 3d scenes also contain a terrain, which is never transparent. However, I have also been working on adding Z rotation support for the extent of a 3d scene (I will also create a PR for this feature). In that case, the drawing of the texture of a terrain becomes more complicated because of the intersection between the rotated 3d scene extent and the unrotated bounding box of the terrain. After some discussion with @uclaros last year, we came to the conclusion that the best way to support it was to make the terrain transparent in order  to hide parts of the terrain which do not need to be drawn. This means that the current state of the camera controller in that case. Hence, this PR.

This issue is fixed by using a Qt3D ray caster to detect the clicked point + using the camera distance in
`Qgs3DUtils::screenPointToWorldPos` if the clicked point is not an entity.

uclaros commented:
I remember having issues with the Qt3D ray caster when doing the 3d measure tool rework and eventually not using it. One major issue is pointclouds. While it _might_ work if you use `Qt3DRender::QPickingSettings::PrimitivePicking` instead of triangles, still that would give a terrible ux. For camera navigation we do not want to rely on a ultra precise ray casting to get the depth since that would mean you would get no hits when the cursor is in-between points, even if the points are quite dense at that location.

I'd try to look for an alternative solution. Why are the transparent objects not in the depth buffer? Is that required for rendering? Can't we just add them? Even if not, I'd rather add a second depth buffer that's only used for camera control including transparent objects than going the Qt3D rayCaster way.

ptitjano commented:
> > Why are the transparent objects not in the depth buffer? Is that required for rendering?
> 
> In the render pass for semi-transparent objects, we disable writes to depth buffer - this is to make sure that it does not matter in which order the semi-transparent triangles are being rendered.
> 
> If we need support for semi-transparent objects, the semi-transparent rendering pass could be updated to also write to the depth buffer (there are probably more ways to do it).

Thanks for your comments @uclaros and @wonder-sk .  
@wonder-sk could you indicate which parts of the semi-transparent rendering pass need to be updated because I initially tried to do so but I have been unable to make it work.

wonder-sk commented:
@ptitjano what did you try that did not work?

I would probably try to implement it by doing semi-transparent rendering in two passes: first pass keep as is, then do a second pass with color output disabled and depth writes enabled. There are maybe better ways to do it, but this should be quite simple...

ptitjano commented:
> color output disabled



> @ptitjano what did you try that did not work?
> 
> I would probably try to implement it by doing semi-transparent rendering in two passes: first pass keep as is, then do a second pass with color output disabled and depth writes enabled. There are maybe better ways to do it, but this should be quite simple...

Thanks for the reponse. This approach but I think this comes from this change from @nyalldawson: https://github.com/qgis/QGIS/pull/55739/commits/063307ecd144c86834d3e6770746abaf79e8404e

Removing the "Qt3DRender::QNoDepthMask" in the material makes it work. Do you think this is good idea?

ptitjano committed:
phongmaterialsettings: Do not add redundant renderpass parameters

There is no need for these parameters because they are already
defined in the transparent filter of the framegraph.

ptitjano committed:
qgsframegraph: Add semi-transparent depth write

QgsCameraController uses the depth buffer of the framegraph to get the
depth when clicking on the 3d view for a rotation or a zoom
operation. This works very well for opaque objects. However, this does
not work for transparent objects because they are not written to the
depth buffer.

This issue is fixed by adding a third branch to the forward render
pass. This new branch only writes the depth of transparent
objets. This allows to have transparent objects written in the depth
buffer and solves the issue with the camera controller.

wonder-sk commented:
> Removing the "Qt3DRender::QNoDepthMask" in the material makes it work. Do you think this is good idea?

Not really, it would be a good idea only if we had all semi-transparent triangles always ordered back-to-front (which would be expensive)

nyalldawson commented:
@ptitjano the new test failures are valid -- they are a result of f5a144f, which has (re)broken transparency support :smile:

nyalldawson commented:
LGTM! @wonder-sk are you ok with this version?

ptitjano commented:
> > Removing the "Qt3DRender::QNoDepthMask" in the material makes it work. Do you think this is good idea?
> 
> Not really, it would be a good idea only if we had all semi-transparent triangles always ordered back-to-front (which would be expensive)

Thanks for your comment. However, I have talked with a Qt3D developer from KDAB and he explained to me that there is no need to define it in the material because it is already defined in the framegraph.

wonder-sk commented:
> Thanks for your comment. However, I have talked with a Qt3D developer from KDAB and he explained to me that there is no need to define it in the material because it is already defined in the framegraph.

There's some confusion I guess - what I meant is that without QNoDepthMask render state specified at all, the semi-transparent rendering would not work properly.

The advice from KDAB sounds correct, and I think it should be also safe to remove `QBlendEquation` from `QgsPhongMaterialSettings`, because it is already set in the frame graph.

If you are up for more cleanup in `QgsPhongMaterialSettings`, I think all of the opacity related code there that you wanted to remove _could_ be actually removed. The code is apparently there only because `QgsFrameGraph` has incorrect `QBlendEquationArguments` setup, and instead it should be set up the way as `QgsPhongMaterialSettings` sets it:
```cpp
    blendState->setSourceRgb( Qt3DRender::QBlendEquationArguments::SourceAlpha );
    blendState->setDestinationRgb( Qt3DRender::QBlendEquationArguments::OneMinusSourceAlpha );
```
(source alpha and destination alpha can stay with default values - `One` for source and `Zero` for destination)

If you fix this in the frame graph's transparent branch, then hopefully the render state setup in `QgsPhongMaterialSettings` should not be needed, and it will make the code cleaner.

But even without this change, the code looks OK to me so good to merge...

ptitjano committed:
qgsframegraph: Fix blend equation args for semitransparent entities

See discussion from
https://github.com/qgis/QGIS/pull/55943#issuecomment-1972177646

ptitjano commented:
> > Thanks for your comment. However, I have talked with a Qt3D developer from KDAB and he explained to me that there is no need to define it in the material because it is already defined in the framegraph.
> 
> There's some confusion I guess - what I meant is that without QNoDepthMask render state specified at all, the semi-transparent rendering would not work properly.
> 
> The advice from KDAB sounds correct, and I think it should be also safe to remove `QBlendEquation` from `QgsPhongMaterialSettings`, because it is already set in the frame graph.
> 
> If you are up for more cleanup in `QgsPhongMaterialSettings`, I think all of the opacity related code there that you wanted to remove _could_ be actually removed. The code is apparently there only because `QgsFrameGraph` has incorrect `QBlendEquationArguments` setup, and instead it should be set up the way as `QgsPhongMaterialSettings` sets it:
> 
> ```c++
>     blendState->setSourceRgb( Qt3DRender::QBlendEquationArguments::SourceAlpha );
>     blendState->setDestinationRgb( Qt3DRender::QBlendEquationArguments::OneMinusSourceAlpha );
> ```
> 
> (source alpha and destination alpha can stay with default values - `One` for source and `Zero` for destination)
> 
> If you fix this in the frame graph's transparent branch, then hopefully the render state setup in `QgsPhongMaterialSettings` should not be needed, and it will make the code cleaner.
> 
> But even without this change, the code looks OK to me so good to merge...

Thanks for the explanation. This confirms what I thought. I have updated the PR to fix the `blendEquation` issues.
